ergo/distributions/point_density.py

```diff
@@ -0,0 +1,199 @@
+        return (class_params, numeric_params)
+
+    # Optimizable
+
+    @classmethod
+    def from_conditions(cls, *args, fixed_params=None, scale=None, **kwargs):
+        if scale is None:
+            # TODO: Should we do this?
+            scale = Scale(0, 1)
+        if fixed_params is None:
+            fixed_params = {
+                "xs": np.linspace(
+                    scale.low, scale.high, constants.point_density_default_num_points
+                )
+            }
+        return super(PointDensity, cls).from_conditions(
+            *args, fixed_params=fixed_params, scale=scale, **kwargs
+        )
+
+    @classmethod
+    def from_params(cls, fixed_params, opt_params, scale=None, traceable=True):
+        # TODO: traceable is always True here
+        if scale is None:
+            # TODO: Should we do this?
+            scale = Scale(0, 1)
+        xs = fixed_params["xs"]
+        ps = np.abs(opt_params)
+        Z = np.sum(cls.normed_bin_probs(xs, ps))
+        densities = ps / Z
+        # print(np.sum(cls.normed_bin_probs(xs, densities)))
+        return cls(
+            xs=xs, densities=densities, scale=scale, normalized=True, traceable=True
+        )
+
+    @staticmethod
+    def initialize_optimizable_params(fixed_params):
+        num_points = fixed_params["xs"].size
+        return onp.full(num_points, 1.0)
+
+    @classmethod
+    def normalize_fixed_params(self, fixed_params, scale):
+        return {"xs": scale.normalize_points(fixed_params["xs"])}
+
+    # Other
+
+    @classmethod
+    def from_pairs(cls, pairs, scale: Scale, normalized=False):
+        sorted_pairs = sorted([(v["x"], v["density"]) for v in pairs])
+        xs = [x for (x, density) in sorted_pairs]
+        densities = [density for (x, density) in sorted_pairs]
+        return cls(xs, densities, scale=scale, normalized=normalized)
+
+    def to_lists(self):
+        xs = self.scale.denormalize_points(self.normed_xs)
+        densities = self.scale.denormalize_densities(self.xs, self.normed_densities)
+        return xs, densities
+
+    def to_pairs(self):
+        xs, densities = self.to_lists()
+        pairs = [
+            {"x": float(x), "density": float(density)}
+            for x, density in zip(xs, densities)
+        ]
+        return pairs
+
+    def to_arrays(self):
+        xs, densities = self.to_lists()
+        return np.array(xs), np.array(densities)
+
+    def entropy(self):
+        # We assume that the distributions are on the same scale!
+        return -np.dot(self.normed_densities, self.normed_log_densities)
+
+    def cross_entropy(self, q_dist):
+        # We assume that the distributions are on the same scale!
+        return -np.dot(self.normed_densities, q_dist.normed_log_densities)
+
+    @classmethod
+    def normed_bin_probs(cls, normed_xs, normed_densities, numpy=np):
+        bin_sizes = numpy.diff(normed_xs)
+        bin_probs = (normed_densities[1:] + normed_densities[:-1]) / 2.0 * bin_sizes
```

Comment:
We're taking the probability at the bin mean value here. Not sure this is compatible with what we're doing elsewhere.